core/rawdb/accessors_metadata.go

```diff
@@ -79,3 +80,60 @@ func WriteChainConfig(db ethdb.KeyValueWriter, hash common.Hash, cfg *params.Cha
 		log.Crit("Failed to store chain config", "err", err)
 	}
 }
+
+// crashList is a list of unclean-shutdown-markers, for rlp-encoding to the
+// database
+type crashList struct {
+	Discarded uint64   // how many ucs have we deleted
+	Recent    []uint64 // unix timestamps of 10 latest unclean shutdowns
+}
+
+const crashesToKeep = 10
+
+// PushUncleanShutdownMarker appends a new unclean shutdown marker and returns
+// the previous data
+// - a list of timestamps
+// - a count of how many old unclean-shutdowns have been discarded
+func PushUncleanShutdownMarker(db ethdb.KeyValueStore) ([]uint64, uint64, error) {
+	var uncleanShutdowns crashList
+	// Read old data
+	if data, err := db.Get(uncleanShutdownKey); err != nil {
+		log.Warn("Error reading unclean shutdown markers", "error", err)
+	} else if err := rlp.DecodeBytes(data, &uncleanShutdowns); err != nil {
+		return nil, 0, err
+	}
+	var discarded = uncleanShutdowns.Discarded
+	var previous = make([]uint64, len(uncleanShutdowns.Recent))
+	copy(previous, uncleanShutdowns.Recent)
+	// Add a new (but cap it)
+	uncleanShutdowns.Recent = append(uncleanShutdowns.Recent, uint64(time.Now().Unix()))
+	if l := len(uncleanShutdowns.Recent); l > crashesToKeep+1 {
+		uncleanShutdowns.Recent = uncleanShutdowns.Recent[l-crashesToKeep-1:]
```

Comment:
Thats a pretty fancy way to discard the first element, wouldn't `Recent[1:]` suffice?
(I assume that only one element can be added/discarded, since `Discarded++` is not `Discarded += len(Recents)-crashesToKeep`)

Reply:
Myeah, I intended for it to handle the case where we discard more than one, e.g. if we decide to change the limit to `5` or there's some mishap. 
So the proper thing to do is rather to uncrease `uncleanShutdowns.Discarded` appropriately. Good catch!